### PR TITLE
Add isSecurePaymentConfirmationAvailable API to spec

### LINKED ISSUE #81: Make it easy to determine whether a user agent supports SPC
Based on conversations with developers, there is a clear desire for an easy mechanism to know whether a browser/user agent supports SPC.

(Related: Issue #65.)

### PR DESCRIPTION
**Pull request by nickburris**
This adds spec text for an API for developers to more easily check whether SPC is available. Fixes #81. One question is whether the spec should further clarify what it means for SPC to be available, but I'm not sure how we can say that without being too implementation specific.

***
<a href="https://pr-preview.s3.amazonaws.com/nickburris/secure-payment-confirmation/pull/233.html" title="Last updated on Apr 11, 2023, 7:31 PM UTC (484e07a)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/secure-payment-confirmation/233/568fbab...nickburris:484e07a.html" title="Last updated on Apr 11, 2023, 7:31 PM UTC (484e07a)">Diff</a>

### CODE REVIEW

**Comment by ianbjacobs** [comment]
Thanks @nickburris and @stephenmcgruer, 

I don't have a strong view; let's try to get input from users of the API. 

Having said that, I see some small advantages to isSecurePaymentConfirmationAvailable:

1) isSecurePaymentConfirmationAvailable appears to be more straightforward in meaning than canMakePayment.
2) This new approach might ease a transition away from SPC-via-Payment-Request if we choose that path.

**Comment by stephenmcgruer** [comment]
> I don't have a strong view; let's try to get input from users of the API.

How/when do you want to gather this input?

> Having said that, I see some small advantages to isSecurePaymentConfirmationAvailable: [...]

Yes, I think we definitely want to make isSecurePaymentConfirmationAvailable the primary way to do availability-checking for the API. The question is more "what about canMakePayment" in such a world - should the SPC spec try to address it (either by defining it to do 'something' useful, TBD what, or by explicitly saying not to use it for SPC)?

**Comment by ianbjacobs** [comment]
> > I don't have a strong view; let's try to get input from users of the API.
> 
> How/when do you want to gather this input?

I have reached out to Stripe and Adyen colleagues for input. I'll also add this to our next meeting agenda.

> 
> > Having said that, I see some small advantages to isSecurePaymentConfirmationAvailable: [...]
> 
> Yes, I think we definitely want to make isSecurePaymentConfirmationAvailable the primary way to do availability-checking for the API. The question is more "what about canMakePayment" in such a world - should the SPC spec try to address it (either by defining it to do 'something' useful, TBD what, or by explicitly saying not to use it for SPC)?

While SPC is grounded in PR API I don't see a reason to try to get rid of canMakePayment. So would this make sense language for users of the API:

* Use isSecurePaymentConfirmationAvailable
* canMakePayment is the old way of doing this; this approach is deprecated.
* Implementations will continue to support both (until the mythical day when SPC is no longer grounded in PR API, at which point canMakePayment could naturally fall away).

**Comment by stephenmcgruer** [comment]
> I have reached out to Stripe and Adyen colleagues for input. I'll also add this to our next meeting agenda.

Ack. Note that we would like to ship this in M113 (branches March 23rd, for stable release in early May), so we are likely to pursue that anyway as we don't foresee any objections from the WG. If for some reason there is significant push-back at the remote meeting (week of March 27th), we should be fine to halt the brakes that close post-branch.

> While SPC is grounded in PR API I don't see a reason to try to get rid of canMakePayment. So would this make sense language for users of the API:

Yep, sgtm. Practically, what I believe this amounts to is:

1. Spec isSecurePaymentConfirmationAvailable
2. Add a spec note saying "canMakePayment also exists, but we recommend using isSecurePaymentConfirmationAvailable instead"

Does that SGTY @ianbjacobs ?

**Comment by omertoast** [comment]
> 1. Spec isSecurePaymentConfirmationAvailable
> 2. Add a spec note saying "canMakePayment also exists, but we recommend using isSecurePaymentConfirmationAvailable instead"

SGTM. As someone who didn't know this issue exist, it seems straightforward and simple as it gets.

**Comment by adrianhopebailie** [comment]
> until the mythical day when SPC is no longer grounded in PR API

`< sigh />`

**Comment by nickburris** [comment]
> Can you update the example in section 1.2.2 to: (1) use your `const spcAvailable = ...`, and to (2) remove or update the comment about canMakePayment ?

Done.

File: spec.bs
Comment (stephenmcgruer) [inline spec.bs:406]: Optional: I wonder if we should move this up to line 366, where there is already a feature detection check?

File: spec.bs
Comment (stephenmcgruer) [inline spec.bs:653]: Lets add something at the bottom like the following (but maybe with better wording than me :D):

```
NOTE: The use of the static {{PaymentRequest/isSecurePaymentConfirmationAvailable}} method is recommended for
      SPC feature detection, instead of calling {{PaymentRequest/canMakePayment}} on an already-constructed
      PaymentRequest object.
```

**Review by stephenmcgruer — APPROVED** [review]
LGTM with one comment and one optional comment. Thanks Nick!

File: spec.bs
Comment (nickburris) [reply spec.bs:656]: Done, that wording SGTM!

**Review by stephenmcgruer — APPROVED** [review]

**Comment by stephenmcgruer** [comment]
@ianbjacobs - PTAL as well. Let us know also if you want this to wait for a WPWG meeting to let anyone chime in, or if you're happy to see it merged after your review.

File: spec.bs
Comment (ianbjacobs) [inline spec.bs:640]: Is there a case where the UA would believe it is supported but it is not supported? If not, can we say "if the Secure Payment Confirmation feature is available" ?

Comment (stephenmcgruer) [reply spec.bs:640]: Removing 'believe' SGTM. If we look at ConditionalUI:

https://w3c.github.io/webauthn/#dom-publickeycredential-isconditionalmediationavailable

```
[PublicKeyCredential](https://w3c.github.io/webauthn/#publickeycredential) overrides this method to indicate availability for [conditional](https://w3c.github.io/webappsec-credential-management/#dom-credentialmediationrequirement-conditional) mediation. [WebAuthn Relying Parties](https://w3c.github.io/webauthn/#webauthn-relying-party) SHOULD verify availability before attempting to set options.[mediation](https://w3c.github.io/webappsec-credential-management/#dom-credentialrequestoptions-mediation) to [conditional](https://w3c.github.io/webappsec-credential-management/#dom-credentialmediationrequirement-conditional).

Upon invocation, a promise is returned that resolves with a value of true if [conditional](https://w3c.github.io/webappsec-credential-management/#dom-credentialmediationrequirement-conditional) [user mediation](https://w3c.github.io/webappsec-credential-management/#user-mediated) is available, or false otherwise.

This method has no arguments and returns a promise to a Boolean value.
```

So they just say 'is available'.

**Review by ianbjacobs — APPROVED** [review]

**Comment by ianbjacobs** [comment]
Hi @nickburris,

I've approved the request, but this should not be part of the document that we advance to Candidate Recommendation. I do expect us to fold in other pull requests that will be part of the CR (namely, #230, #234, #236, and #238). Should we block merging this for now?

**Comment by stephenmcgruer** [comment]
> Should we block merging this for now?

Yep; let's just leave this PR sitting until the first version of SPC gets to CR (assuming no significant bumps along the way). On the Chrome side, an open approved PR is sufficient for our launch processes :)

**Comment by RByers** [comment]
[Chromium API owner hat on] Do y'all have an idea of how long you expect SPC to take to get to CR? While we can ship based on a PR if necessary, it's not ideal as it can contribute to confusion about what's ready to be implemented in other engines. If you expect it to take more than a month or so, could you consider creating a separate 'v2' branch or something which we can point to as the landed spec for this API?

**Comment by ianbjacobs** [comment]
Hi @RByers,
We are poised to request to advance to CR. I am hopeful for a mid-May publication.

**Comment by ianbjacobs** [comment]
@RByers, yes, very open to a versioning / branch discussion.

**Comment by ianbjacobs** [comment]
@stephenmcgruer,
I'm ok to land the pull request, but would prefer to do so after the publication of the CR (ideally, therefore, in a few days). Would it be ok to wait for that? If you want to delegate landing this to me (for just after the publication), I'm happy to merge it.

**Comment by stephenmcgruer** [comment]
Sure, sounds good to me - if you can take the task to land this once the CR is published, SGTM. Thanks! :)

**Pull request merged**
